### PR TITLE
ref(*): use the kubernetes client directly

### LINKED ISSUE #65: Fix race condition while attempting to stream logs from the builder pod.
https://github.com/deis/builder/blob/daac4b822a94554799c8503e5e7662592492b7f0/rootfs/etc/confd/templates/builder#L139

Is very prone to race conditions. If the builder pod either never makes it to `Running` or transitions through `Running` to `Terminated` we might hang forever.

### PR DESCRIPTION
**Pull request by aledbf**
replaces #115 
closes #65

### CODE REVIEW

**Comment by arschles** [comment]
Replaces https://github.com/deis/builder/pull/115

Also, @aledbf talked offline - he'll be writing tests in this PR

**Comment by arschles** [comment]
Closes #79 
Fixes #65

File: pkg/gitreceive/k8s_util.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	pod := buildPod("slugbuilder", withAuth, name, namespace, env)
+	return &pod
+}
+
+func buildPod(buildType string, withAuth bool, name, namespace string, env map[string]interface{}) api.Pod {
+	useCreds := "-no-creds"
+	if withAuth {
+		useCreds = ""
+	}
+	fileName := fmt.Sprintf("/etc/deis-%v%v.yaml", buildType, useCreds)
```
Comment (arschles) [inline pkg/gitreceive/k8s_util.go]: can we completely remove the on-disk templates and instead construct them entirely in memory?

File: pkg/gitreceive/run.go [hunk 2]
```diff
@@ -37,6 +39,8 @@ func Run(conf *Config) error {
 		return fmt.Errorf("configuring S3 client (%s)", err)
 	}
 
+	kubeClient, err := client.NewInCluster()
```
Comment (arschles) [inline pkg/gitreceive/run.go]: need to check the `err` here and `return` it if it's non-nil

File: pkg/gitreceive/build.go [hunk 3]
```diff
@@ -183,62 +138,94 @@ func build(conf *Config, s3Client *s3.S3, builderKey, gitSha string) error {
...
+		case <-timeout:
+			return fmt.Errorf("Timed out waiting for object in storage. Aborting build...")
+		case <-tick:
+			exists, err := storage.ObjectExists(s3Client, bucketName, pushObjKey)
+			if err != nil {
+				return fmt.Errorf("Checking if object %s/%s exists (%s)", bucketName, pushObjKey, err)
+			}
+			if exists {
+				break
+			}
```
Comment (arschles) [inline pkg/gitreceive/build.go]: I like this, but it seems out-of-scope for this patch. Can you move it into a different pull request?

File: pkg/gitreceive/build.go [hunk 4]
```diff
@@ -183,62 +138,94 @@ func build(conf *Config, s3Client *s3.S3, builderKey, gitSha string) error {
...
-		"create",
-		"-f",
-		finalManifestFileLocation,
-	)
-	if log.IsDebugging {
-		kCreateCmd.Stdout = os.Stdout
-	}
-	kCreateCmd.Stderr = os.Stderr
-	if err := run(kCreateCmd); err != nil {
+	log.Debug("Pod spec: %v", pod)
```
Comment (arschles) [inline pkg/gitreceive/build.go]: Will this output something reasonably human-readable? If not, can you write something that will generate a cleaner-looking string?

File: pkg/gitreceive/build.go [hunk 5]
```diff
@@ -14,8 +14,13 @@ import (
...
 	"github.com/deis/builder/pkg/gitreceive/log"
 	"github.com/deis/builder/pkg/gitreceive/storage"
-	"github.com/pborman/uuid"
 	"gopkg.in/yaml.v2"
+
+	"k8s.io/kubernetes/pkg/api"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/watch"
```
Comment (arschles) [inline pkg/gitreceive/build.go]: I don't see these dependencies explicitly defined in the `glide.yaml` file, and I believe that they're a transitive dependency from `github.com/deis/pkg`. That package uses version `~1.1` (https://github.com/deis/pkg/blob/master/glide.yaml#L4-L5), which is indeed our target for now, but I'd like to explicitly define the version in this repository in case the `deis/pkg` changes it in the future.

File: pkg/gitreceive/build.go [hunk 6]
```diff
@@ -169,51 +125,77 @@ func build(conf *Config, s3Client *s3.S3, builderKey, rawGitSha string) error {
...
+	}
+
+	ch := watcher.ResultChan()
+	for evt := range ch {
+		if evt.Type == watch.Added {
+			watcher.Stop()
 			break
-		} else if strings.Contains(outStr, "phase: Failed") {
-			return fmt.Errorf("build pod %s entered phase: Failed", buildPodName)
+		} else if evt.Type == watch.Error {
```
Comment (arschles) [inline pkg/gitreceive/build.go]: is there a `watch.Failed` that we need to look for?

File: pkg/gitreceive/build.go [hunk 7]
```diff
@@ -169,51 +125,77 @@ func build(conf *Config, s3Client *s3.S3, builderKey, rawGitSha string) error {
...
-	// TODO: use the k8s client and watch the event stream instead (https://github.com/deis/builder/issues/65)
-	for {
-		cmd := kGetCmd(conf.PodNamespace, buildPodName)
-		var out bytes.Buffer
-		cmd.Stdout = &out
-		// ignore errors
-		run(cmd)
-		outStr := string(out.Bytes())
-		if strings.Contains(outStr, "phase: Running") {
+	watcher, err := podsInterface.Watch(labels.Nothing(), fields.OneTermEqualSelector("name", pod.Name), v1Version)
```
Comment (arschles) [inline pkg/gitreceive/build.go]: If the pod has finished starting and is running by the time this call occurs, wouldn't the watch return no events?

File: pkg/gitreceive/build.go [hunk 8]
```diff
@@ -265,3 +247,15 @@ func build(conf *Config, s3Client *s3.S3, builderKey, rawGitSha string) error {
 
 	return nil
 }
+
+func prettyPrintJSON(data interface{}) string {
+	output := &bytes.Buffer{}
+	if err := json.NewEncoder(output).Encode(data); err != nil {
+		return ""
```
Comment (arschles) [inline pkg/gitreceive/build.go]: This should return the error instead of suppressing it

File: pkg/gitreceive/build.go [hunk 9]
```diff
@@ -265,3 +247,15 @@ func build(conf *Config, s3Client *s3.S3, builderKey, rawGitSha string) error {
...
+	}
+	formatted := &bytes.Buffer{}
+	if err := json.Indent(formatted, output.Bytes(), "", "  "); err != nil {
+		return ""
```
Comment (arschles) [inline pkg/gitreceive/build.go]: Same here regarding returning the error

File: pkg/gitreceive/build.go (hunk 6 above)
Comment (aledbf) [inline pkg/gitreceive/build.go]: that comes from your PR 115 ([here](https://github.com/deis/builder/pull/115/files#diff-91e10a0836e513c85626cee1d2899e5dR208))

File: pkg/gitreceive/k8s_util.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+	pod.ObjectMeta.Name = name
+	pod.ObjectMeta.Namespace = namespace
+
+	return pod
+}
+
+func podFromFile(fileName string) api.Pod {
+	var pod api.Pod
+	data, err := ioutil.ReadFile(fileName)
```
Comment (arschles) [inline pkg/gitreceive/k8s_util.go]: Same comment here regarding removing the file completely. Ideally, the files would be completely removed from the repository (and hence, not built into the resulting docker image) in this patch.

File: pkg/gitreceive/k8s_util_test.go [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+	}
+}
+
+func buildTestPod(name, namespace string, env map[string]interface{}, templ string) api.Pod {
+	pod := podFromFile(templ)
+	pod.ObjectMeta.Name = name
+	pod.ObjectMeta.Namespace = namespace
+	addEnvToPod(pod, env)
+	return pod
+}
```
Comment (arschles) [inline pkg/gitreceive/k8s_util_test.go]: :100: for these tests!

File: pkg/gitreceive/build.go (hunk 6 above)
Comment (arschles) [inline pkg/gitreceive/build.go]: Makes sense. Do you happen to know if we also need to look for a `Failed` event?

Comment (aledbf) [inline pkg/gitreceive/build.go]: No. Check [here]( https://github.com/kubernetes/kubernetes/blob/458eb5284abd6933c620ba8ea383032f225e0885/pkg/watch/watch.go#L40) for the options

File: pkg/gitreceive/build.go (hunk 7 above)
Comment (aledbf) [inline pkg/gitreceive/build.go]: Good point. I added resourceVersion to indicate what version should be used to begin watching

File: pkg/gitreceive/build.go (hunk 6 above)
Comment (arschles) [inline pkg/gitreceive/build.go]: Got it. I think it would be prudent to also check for `Deleted` as well, even though I doubt it would come up. Thoughts?

File: pkg/gitreceive/build.go [hunk 12]
```diff
@@ -169,51 +125,82 @@ func build(conf *Config, s3Client *s3.S3, builderKey, rawGitSha string) error {
...
-	// TODO: use the k8s client and watch the event stream instead (https://github.com/deis/builder/issues/65)
-	for {
-		cmd := kGetCmd(conf.PodNamespace, buildPodName)
-		var out bytes.Buffer
-		cmd.Stdout = &out
-		// ignore errors
-		run(cmd)
-		outStr := string(out.Bytes())
-		if strings.Contains(outStr, "phase: Running") {
+	watcher, err := podsInterface.Watch(labels.Nothing(), fields.OneTermEqualSelector("name", pod.Name), newPod.ResourceVersion)
```
Comment (arschles) [inline pkg/gitreceive/build.go]: The godoc I see in code isn't entirely clear about the `ResourceVersion`. Does calling  effectively allow you to get events from the beginning of the pod's lifetime?

File: pkg/gitreceive/k8s_util.go [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+	if err != nil {
+		fmt.Errorf("invalid pod template %v (%v)", fileName, err)
+	}
+
+	api.Scheme.DecodeInto(json, &pod)
+
+	return pod
+}
+
+func addEnvToPod(pod api.Pod, env map[string]interface{}) {
```
Comment (arschles) [inline pkg/gitreceive/k8s_util.go]: why not make `env` a `map[string]string`?

**Comment by arschles** [comment]
@aledbf I left a few small comments just now. My big question still remaining is - can you remove the manifest files and generate them purely from memory?

File: pkg/gitreceive/k8s_util.go (hunk 13 above)
Comment (aledbf) [inline pkg/gitreceive/k8s_util.go]: No particular reason. Just using [this](https://github.com/deis/builder/blob/master/pkg/types.go#L38)

Comment (arschles) [inline pkg/gitreceive/k8s_util.go]: I should have read up! Fine with me as is, then.

**Comment by aledbf** [comment]
@arschles here is the output from a push
```
$ git push deis master
Counting objects: 6, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (3/3), done.
Writing objects: 100% (6/6), 511 bytes | 0 bytes/s, done.
Total 6 (delta 0), reused 0 (delta 0)
Running in debug mode
Running git hook
read [0000000000000000000000000000000000000000,75aeeee1695a35d785a49204474158e7eaf5890d,refs/heads/master]
Workflow request /v2/hooks/push (body elided)
Workflow request POST /v2/hooks/config
{"receive_user":"alejandro.debrito","receive_repo":"issue124"}
got the following config back for app issue124: {Owner:alejandro.debrito App:issue124 Values:map[] Memory:map[] CPU:map[] Tags:map[] UUID:afc70872-de58-4ce3-b692-ad84e2039509 Created:2016-01-25 21:58:49 +0000 UTC Updated:2016-01-25 21:58:49 +0000 UTC}
running [git archive --format=tar.gz --output=issue124.tar.gz 75aeeee1695a35d785a49204474158e7eaf5890d] in directory /home/git/issue124.git
running [tar -xzf issue124.tar.gz -C /tmp/] in directory /home/git/issue124.git
create bucket error: BucketAlreadyExists: The requested bucket name is not available.
remote: status code: 409, request id: 790FXO784V6IADJX
Uploading tar to http://10.3.3.23:9000/git/home/issue124:git-75aeeee1/tar
-----> Starting build... but first, coffee!
Starting pod slugbuild-issue124-75aeeee1-e692d8c3
Pod spec: {
  "metadata": {
    "name": "slugbuild-issue124-75aeeee1-e692d8c3",
    "namespace": "deis",
    "creationTimestamp": null,
    "labels": {
      "heritage": "deis",
      "version": "2.0.0-beta"
    }
  },
  "spec": {
    "volumes": [
      {
        "name": "minio-user",
        "secret": {
          "secretName": "minio-user"
        }
      }
    ],
    "containers": [
      {
        "name": "deis-slugbuilder",
        "image": "quay.io/deisci/slugbuilder:v2-beta",
        "env": [
          {
            "name": "TAR_URL",
            "value": "http://10.3.3.23:9000/git/home/issue124:git-75aeeee1/tar"
          },
          {
            "name": "put_url",
            "value": "http://10.3.3.23:9000/git/home/issue124:git-75aeeee1695a35d785a49204474158e7eaf5890d/push"
          }
        ],
        "resources": {},
        "volumeMounts": [
          {
            "name": "minio-user",
            "readOnly": true,
            "mountPath": "/var/run/secrets/object/store"
          }
        ],
        "imagePullPolicy": "Always"
      }
    ],
    "restartPolicy": "Never",
    "serviceAccountName": ""
  },
  "status": {}
}

-----> Unable to select a buildpack ---> (this is "expected", there's only a Procfile file in the directory)
size of logs streamed 40
-----> Build complete.
-----> Launching app.
-----> Launching...
Workflow request POST /v2/hooks/build
{"sha":"75aeeee1695a35d785a49204474158e7eaf5890d","receive_user":"alejandro.debrito","receive_repo":"issue124","image":"issue124","procfile":{"web":"true"},"dockerfile":""}
[DEBUG]
-----> Done, issue124:v6 deployed to Deis

-----> Use 'deis open' to view this application in your browser

-----> To learn more, use 'deis help' or visit http://deis.io

running [git gc] in directory /home/git/issue124.git
To ssh://git@deis-v2:2222/issue124.git
 * [new branch]      master -> master
```

File: pkg/gitreceive/build.go [hunk 14]
```diff
@@ -169,52 +123,75 @@ func build(conf *Config, s3Client *s3.S3, builderKey, rawGitSha string) error {
...
-		outStr := string(out.Bytes())
-		if strings.Contains(outStr, "phase: Running") {
-			break
-		} else if strings.Contains(outStr, "phase: Failed") {
-			return fmt.Errorf("build pod %s entered phase: Failed", buildPodName)
-		}
-		time.Sleep(100 * time.Millisecond)
+	timeout := time.Duration(5 * time.Minute)
+	tick := time.Duration(500 * time.Millisecond)
+	waitForPod(kubeClient, newPod.Namespace, newPod.Name, tick, timeout)
```
Comment (arschles) [inline pkg/gitreceive/build.go]: can you merge this line and the below, like so:

```console
if err := waitForPod(kubeClient, newPod.Namespace, newPod.Name, tick, timeout); err != nil {
  ...
}
```

Comment (aledbf) [inline pkg/gitreceive/build.go]: done

File: pkg/gitreceive/build.go [hunk 15]
```diff
@@ -169,52 +123,74 @@ func build(conf *Config, s3Client *s3.S3, builderKey, rawGitSha string) error {
...
+		log.Debug("Pod spec: %v", json)
+	} else {
+		log.Debug("Error creating json representaion of pod spec: %v", err)
 	}
-	kCreateCmd.Stderr = os.Stderr
-	if err := run(kCreateCmd); err != nil {
+
+	podsInterface := kubeClient.Pods(conf.PodNamespace)
+
+	//creationTime := unversioned.Now()
```
Comment (arschles) [inline pkg/gitreceive/build.go]: delete this commented line?

File: pkg/gitreceive/k8s_util.go [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+func slugBuilderPodName(appName, shortSha string) string {
+	uid := uuid.New()[:8]
+	return fmt.Sprintf("slugbuild-%s-%s-%s", appName, shortSha, uid)
+}
+
+func dockerBuilderPod(debug, withAuth bool, name, namespace string, env map[string]interface{}, tarURL, imageName string) *api.Pod {
+	pod := buildPod(debug, withAuth, name, namespace, env)
+
+	pod.Spec.Containers[0].Name = "deis-dockerbuilder"
+	pod.Spec.Containers[0].Image = "quay.io/deisci/dockerbuilder:v2-beta"
```
Comment (arschles) [inline pkg/gitreceive/k8s_util.go]: Note to others: a later PR can make this image configurable, and can hence solve https://github.com/deis/builder/issues/66

Comment (arschles) [inline pkg/gitreceive/k8s_util.go]: @aledbf can you put these two and the below two strings in constants for now?

File: pkg/gitreceive/k8s_util.go [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+
+	addEnvToPod(pod, "ACCESS_KEY_FILE", "/var/run/secrets/object/store/access_key")
+	addEnvToPod(pod, "ACCESS_SECRET_FILE", "/var/run/secrets/object/store/access_secret")
+
+	if !withAuth {
+		addEnvToPod(pod, "TAR_URL", tarURL)
+		addEnvToPod(pod, "IMG_NAME", imageName)
```
Comment (arschles) [inline pkg/gitreceive/k8s_util.go]: same here RE constants

File: pkg/gitreceive/k8s_util.go [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+
+	addEnvToPod(pod, "TAR_URL", tarURL)
+	addEnvToPod(pod, "put_url", putURL)
+
+	if buildpackURL != "" {
+		addEnvToPod(pod, "BUILDPACK_URL", buildpackURL)
+	}
+
+	return &pod
+}
```
Comment (arschles) [inline pkg/gitreceive/k8s_util.go]: same here RE the strings in this func

File: pkg/gitreceive/k8s_util.go [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+				Secret: &api.SecretVolumeSource{
+					SecretName: "minio-user",
+				},
+			},
+		})
+
+		pod.Spec.Containers[0].VolumeMounts = []api.VolumeMount{
+			api.VolumeMount{
+				Name:      "minio-user",
+				MountPath: "/var/run/secrets/object/store",
```
Comment (arschles) [inline pkg/gitreceive/k8s_util.go]: same here RE this and the `Name` and `SecretName` strings

File: pkg/gitreceive/k8s_util.go [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+		for k, v := range env {
+			pod.Spec.Containers[0].Env = append(pod.Spec.Containers[0].Env, api.EnvVar{
+				Name:  k,
+				Value: fmt.Sprintf("%v", v),
+			})
+		}
+	}
+
+	if debug {
+		addEnvToPod(pod, "DEBUG", "1")
```
Comment (arschles) [inline pkg/gitreceive/k8s_util.go]: same here RE these two strings

File: pkg/gitreceive/k8s_util_test.go [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+		t.Fatalf("expected pod name dockerbuild-demo-12345678-, got %s", name)
+	}
+}
+
+func SlugBuilderPodName(t *testing.T) {
+	name := dockerBuilderPodName("demo", "12345678")
+	if !strings.HasPrefix(name, "slugbuild-demo-12345678-") {
+		t.Fatalf("expected pod name slugbuild-demo-12345678-, got %s", name)
+	}
+}
```
Comment (arschles) [inline pkg/gitreceive/k8s_util_test.go]: this and the above `func` names need to start with `Test`, or they won't be executed as test cases

**Comment by arschles** [comment]
@aledbf made a few more comments. Also, does this PR fix https://github.com/deis/builder/issues/65?

**Comment by aledbf** [comment]
> Also, does this PR fix #65?

Yes

**Comment by arschles** [comment]
Fixes #65

File: pkg/gitreceive/build.go [hunk 22]
```diff
@@ -169,52 +123,73 @@ func build(conf *Config, s3Client *s3.S3, builderKey, rawGitSha string) error {
...
-		run(cmd)
-		outStr := string(out.Bytes())
-		if strings.Contains(outStr, "phase: Running") {
-			break
-		} else if strings.Contains(outStr, "phase: Failed") {
-			return fmt.Errorf("build pod %s entered phase: Failed", buildPodName)
-		}
-		time.Sleep(100 * time.Millisecond)
+	timeout := time.Duration(5 * time.Minute)
+	tick := time.Duration(500 * time.Millisecond)
```
Comment (arschles) [inline pkg/gitreceive/build.go]: These two values should be configurable. Not blocking this PR. I've created https://github.com/deis/builder/issues/135 to remind.

**Comment by arschles** [comment]
@aledbf I tested on my GKE cluster, and all looks good:

```console
ENG000656:example-go aaronschlesinger$ git push deis master
The authenticity of host '[deis.104.197.73.56.xip.io]:2222 ([104.197.73.56]:2222)' can't be established.
ECDSA key fingerprint is SHA256:pa0axGD3ZJnbdsHCI/dGsWL083DueYSBFMfwjpnSbQ0.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '[deis.104.197.73.56.xip.io]:2222,[104.197.73.56]:2222' (ECDSA) to the list of known hosts.
Counting objects: 115, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (110/110), done.
Writing objects: 100% (115/115), 20.76 KiB | 0 bytes/s, done.
Total 115 (delta 66), reused 0 (delta 0)
-----> Starting build... but first, coffee!
-----> Go app detected
-----> Checking Godeps/Godeps.json file.
-----> Installing go1.4.2... done
-----> Running: godep go install -tags heroku ./...
-----> Discovering process types
       Procfile declares types -> web
-----> Compiled slug size is 1.7M
-----> Build complete.
-----> Launching app.
-----> Launching...
-----> Done, gotest:v2 deployed to Deis

-----> Use 'deis open' to view this application in your browser

-----> To learn more, use 'deis help' or visit http://deis.io

To ssh://git@deis.104.197.73.56.xip.io:2222/gotest.git
 * [new branch]      master -> master
```

**Comment by technosophos** [comment]
This is fantastic! Thanks!
